lib/index.js

```diff
@@ -20,25 +33,71 @@ module.exports = {
       // Step 1: Wait for the arguments to resolve in case they are promises
       B.all([predicateFn, getDispatchFn, valueLens, value])
-        // Step 2: Everything is resolved, compose the two getter-type functions together
-        .then(([predicateFn, getDispatchFn, valueLens, value]) => [
-          valueLens,
-          composeP(asPromise(getDispatchFn), asPromise(predicateFn)),
-          value])
-        // Step 4: Retrieve the dispatch function
-        .then(([valueLens, getDispatch, value]) => B.all([valueLens, getDispatch(value), value, view(valueLens, value)]))
+        // Step 2: Everything is resolved, compose the evaluate the predicateFn
+        .then(([predicateFn, getDispatchFn, valueLens, value]) => {
+          const areFunctions = [predicateFn, getDispatchFn, valueLens].map(is(Function));
+          const failedParams = zip(areFunctions, ['predicateFn', 'getDispatchFn', 'valueLens'])
+            .filter(e => !e[0])
+            .map(e => e[1]);
+
+          if (failedParams.length) {
+            const error = new TypeError(errors.INVALID_FN_PARAMS(failedParams));
+            return handleError(defineTrace, null, error);
+          }
+          if (isNil(value)) {
+            const error = new TypeError(errors.VALUE_NIL());
+            return handleError(defineTrace, null, error);
+          }
+
+          return B.resolve(value)
+            .then(predicateFn)
+            .then((predicateValue) => [valueLens, getDispatchFn, predicateValue, value],
+              handleError(defineTrace, errors.PREDICATE_FN(value)));
+        })
+        // Step 3: Retrieve the dispatch function
+        .then(([valueLens, getDispatchFn, predicateValue, value]) => {
+          return B.resolve(predicateValue)
+            .then(getDispatchFn)
+            .then((dispatcher) => {
+              if (!is(Function, dispatcher)) {
+                const error = new TypeError(errors.DISPATCH_FN_RETURN_VALUE(predicateValue));
+                return handleError(defineTrace, null, error);
+              }
+
+              return B.resolve(value)
+                .then(view(valueLens))
+                .then((focus) => [valueLens, dispatcher, value, focus],
+                  handleError(defineTrace, errors.GET_LENS()));
+            }, handleError(defineTrace, errors.DISPATCH_FN(predicateValue)));
+        })
         // Step 4: Execute the actual dispatch over the data-structure with the provided lens
-        .then(([valueLens, dispatch, value, focus]) => B.all([valueLens, dispatch(focus), value]))
+        .then(([valueLens, dispatcher, value, focus]) => B.all([valueLens, dispatcher(focus), value]))
         // Step 5: Use the return value of the dispatch and the lens to create the real return value
-        .then(([valueLens, returnValue, value]) => set(valueLens, returnValue, value))
+        .then(([valueLens, returnValue, value]) => set(valueLens, returnValue, value),
+          handleError(defineTrace, errors.DISPATCHER_FN()))
```

Comment:
Do you think having the `rejectedHandler` directly in the `.then` chain makes more sense than using the `.catch` sugar?